File: Viewmodel, LiveData/app/src/main/java/junseong/android/myapplication/MainActivity.java
package junseong.android.myapplication;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import java.util.List;

import junseong.android.myapplication.databinding.ActivityMainBinding;
import junseong.android.myapplication.databinding.ItemMainBinding;

public class MainActivity extends AppCompatActivity {


    ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        MyViewModel viewModel= ViewModelProviders.of(this).get(MyViewModel.class);
        viewModel.getNews().observe(this,news -> {
            MyAdapter adapter=new MyAdapter(news);
            binding.recyclerView.setAdapter(adapter);
        });

    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        ItemMainBinding binding;

        public ItemViewHolder(ItemMainBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    class MyAdapter extends RecyclerView.Adapter<ItemViewHolder> {
        List<ItemModel> articles;

        public MyAdapter(List<ItemModel> articles) {
            this.articles = articles;
        }

        @Override
        public int getItemCount() {
            return articles.size();
        }
        @NonNull
        @Override
        public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            ItemMainBinding binding = ItemMainBinding.
                    inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new ItemViewHolder(binding);
        }

        @Override
        public void onBindViewHolder(@NonNull ItemViewHolder holder, int position) {
            ItemModel model = articles.get(position);
            holder.binding.setItem(model);
        }
    }
    @BindingAdapter("bind:publishedAt")
    public static void publishedAt(TextView view, String date) {
        view.setText(AppUtils.getDate(date) + " at " + AppUtils.getTime(date));
    }

    @BindingAdapter("bind:urlToImage")
    public static void urlToImage(ImageView view, String url) {
        Glide.with(MyApplication.getAppContext()).load(url).override(250, 200).into(view);
    }
}

File: RecyclerVIew(kotlin)/app/src/main/java/junseong/kotlin/recyclerview/MainActivity.kt
package junseong.kotlin.recyclerview

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.view.menu.ActionMenuItemView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.row.view.*

class MainActivity : AppCompatActivity() {

    val imgRes= intArrayOf(
            R.drawable.imgflag1,
            R.drawable.imgflag2,
            R.drawable.imgflag3,
            R.drawable.imgflag4,
            R.drawable.imgflag5,
            R.drawable.imgflag6,
            R.drawable.imgflag7,
            R.drawable.imgflag8
            )
    val data1= arrayOf(
            "토고","프랑스","스위스","스페인","일본","독일","브라질","대한민국"
    )
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val adapter1=RecycerAdapter()
        recycler.adapter=adapter1
        /* 리사이클러뷰 세가지방법 리사이클러뷰는 listview와 gridview의 통합형이므로*/

       //recycler.layoutManager=LinearLayoutManager(this)
        //recycler.layoutManager=GridLayoutManager(this,2)
        recycler.layoutManager=StaggeredGridLayoutManager(2,StaggeredGridLayoutManager.VERTICAL)
        //recycler.layoutManager=StaggeredGridLayoutManager(2,StaggeredGridLayoutManager.HORIZONTAL)
    }
    //Recyclerview Adapter class
    inner class RecycerAdapter : RecyclerView.Adapter<RecycerAdapter.ViewHolderClass>(){

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderClass {
            //항목으로 사용할 view객체를 생성
            val itemView=layoutInflater.inflate(R.layout.row,null)
            val holder=ViewHolderClass(itemView)
            itemView.setOnClickListener(holder)
            return holder
        }

        //viewholder를 통해 항목구성할때 항목내의 view객체에 데이터셋팅
        override fun onBindViewHolder(holder: ViewHolderClass, position: Int) {
            holder.rowimageview.setImageResource(imgRes[position])
            holder.rowtext.text=data1[position]
        }

        override fun getItemCount(): Int {
            return imgRes.size
        }

        inner class ViewHolderClass(itemView: View):RecyclerView.ViewHolder(itemView),View.OnClickListener{
            val rowimageview=itemView.rowimage
                val rowtext=itemView.rowtext
            override fun onClick(v: View?) {
                text.text=data1[adapterPosition]
            }
        }

    }
}
